[Feature] Support DeepSeekV3 Function Call

Xu-Wenqing opened the pull request:
Support DeepSeek-V3-0324 function call.

usage:
```
vllm serve ... --enable-auto-tool-choice --tool-call-parser deepseek_v3 --chat-template examples/tool_chat_template_deepseekv3.jinja
```

test script (no streaming):
```Python
from openai import OpenAI

openai_api_base = ""
openai_api_key = ""

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base + "/v1",
)


tools = [
    {
        "type": "function",
        "function": {
            "name": "get_current_temperature",
            "description": "Get current temperature at a location.",
            "parameters": {
                "type": "object",
                "properties": {
                    "location": {
                        "type": "string",
                        "description": 'The location to get the temperature for, in the format "City, State, Country".',
                    },
                    "unit": {
                        "type": "string",
                        "enum": ["celsius", "fahrenheit"],
                        "description": 'The unit to return the temperature in. Defaults to "celsius".',
                    },
                },
                "required": ["location"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "get_temperature_date",
            "description": "Get temperature at a location and date.",
            "parameters": {
                "type": "object",
                "properties": {
                    "location": {
                        "type": "string",
                        "description": 'The location to get the temperature for, in the format "City, State, Country".',
                    },
                    "date": {
                        "type": "string",
                        "description": 'The date to get the temperature for, in the format "Year-Month-Day".',
                    },
                    "unit": {
                        "type": "string",
                        "enum": ["celsius", "fahrenheit"],
                        "description": 'The unit to return the temperature in. Defaults to "celsius".',
                    },
                },
                "required": ["location", "date"],
            },
        },
    },
]


response = client.chat.completions.create(
    model=client.models.list().data[0].id,
    messages=[
        {
            "role": "system",
            "content": "You are a helpful assistant.\n\nCurrent Date: 2024-09-30",
        },
        {
            "role": "user",
            "content": "What's the temperature in San Francisco now? How about tomorrow?",
        },
    ],
    tools=tools,
    tool_choice="auto",
    stream=False,
)

print(response.choices[0].message.content)

tool_calls = response.choices[0].message.tool_calls
for c in tool_calls:
    print(c.function.name, c.function.arguments)
```

Output:
```Bash
None
get_current_temperature {"location": "San Francisco, CA, USA", "unit": "celsius"}
get_temperature_date {"location": "San Francisco, CA, USA", "date": "2024-10-01", "unit": "celsius"}
```


test script (streaming):
```Python
from openai import OpenAI

openai_api_base = ""
openai_api_key = ""

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base + "/v1",
)

class bcolors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKCYAN = '\033[96m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'


tools = [
    {
        "type": "function",
        "function": {
            "name": "get_current_temperature",
            "description": "Get current temperature at a location.",
            "parameters": {
                "type": "object",
                "properties": {
                    "location": {
                        "type": "string",
                        "description": 'The location to get the temperature for, in the format "City, State, Country".',
                    },
                    "unit": {
                        "type": "string",
                        "enum": ["celsius", "fahrenheit"],
                        "description": 'The unit to return the temperature in. Defaults to "celsius".',
                    },
                },
                "required": ["location"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "get_temperature_date",
            "description": "Get temperature at a location and date.",
            "parameters": {
                "type": "object",
                "properties": {
                    "location": {
                        "type": "string",
                        "description": 'The location to get the temperature for, in the format "City, State, Country".',
                    },
                    "date": {
                        "type": "string",
                        "description": 'The date to get the temperature for, in the format "Year-Month-Day".',
                    },
                    "unit": {
                        "type": "string",
                        "enum": ["celsius", "fahrenheit"],
                        "description": 'The unit to return the temperature in. Defaults to "celsius".',
                    },
                },
                "required": ["location", "date"],
            },
        },
    },
]


tool_calls_stream = client.chat.completions.create(
    model=client.models.list().data[0].id,
    messages=[
        {
            "role": "system",
            "content": "You are a helpful assistant.\n\nCurrent Date: 2024-09-30",
        },
        {
            "role": "user",
            "content": "What's the temperature in San Francisco now? How about tomorrow?",
        },
    ],
    tools=tools,
    tool_choice="auto",
    stream=True,
    extra_body={"chat_template_kwargs": {"enable_thinking": True}},
)

print("reasoning content(Blue) and content(Green):")
chunks = []
for chunk in tool_calls_stream:
    chunks.append(chunk)
    if hasattr(chunk.choices[0].delta, "reasoning_content"):
        reasoning_content = chunk.choices[0].delta.reasoning_content
        if reasoning_content:
            print(bcolors.OKBLUE + reasoning_content, end="", flush=True)
    elif hasattr(chunk.choices[0].delta, "content"):
        content = chunk.choices[0].delta.content
        if content:
            print(bcolors.OKGREEN + content, end="", flush=True)

print(bcolors.ENDC + "\n### end of reasoning content and content. ###\n")

arguments = []
tool_call_idx = -1
for chunk in chunks:

    if chunk.choices[0].delta.tool_calls:
        tool_call = chunk.choices[0].delta.tool_calls[0]

        if tool_call.index != tool_call_idx:
            if tool_call_idx >= 0:
                print(f"streamed tool call arguments: {arguments[tool_call_idx]}")
            tool_call_idx = chunk.choices[0].delta.tool_calls[0].index
            arguments.append("")
        if tool_call.id:
            print(f"streamed tool call id: {tool_call.id} ")

        if tool_call.function:
            if tool_call.function.name:
                print(f"streamed tool call name: {tool_call.function.name}")

            if tool_call.function.arguments:
                arguments[tool_call_idx] += tool_call.function.arguments

if len(arguments):
    print(f"streamed tool call arguments: {arguments[-1]}")
```

Output:
```Bash
reasoning content(Blue) and content(Green):


### end of reasoning content and content. ###

streamed tool call id: chatcmpl-tool-e2c23f26c3fa4a45a5eb029babe8ba12 
streamed tool call name: get_current_temperature
streamed tool call arguments: {"location": "San Francisco, USA", "unit": "celsius"}
streamed tool call id: chatcmpl-tool-c4945b99fcbd4134971134c302eaf6a9 
streamed tool call name: get_temperature_date
streamed tool call arguments: {"location": "San Francisco, USA", "date": "2024-10-01", "unit": "celsius"}
```

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 tool parser

Signed-off-by: 许文卿 <xwq391974@alibaba-inc.com>

Xu-Wenqing committed:
Merge branch 'main' into dev/dpsk_r1_tool_parser

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 tool parser

Signed-off-by: 许文卿 <xwq391974@alibaba-inc.com>

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 tool parser

Signed-off-by: 许文卿 <xwq391974@alibaba-inc.com>

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 tool parser

Signed-off-by: 许文卿 <xwq391974@alibaba-inc.com>

Xu-Wenqing commented:
https://github.com/vllm-project/vllm/issues/14745

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 tool parser

Signed-off-by: 许文卿 <xwq391974@alibaba-inc.com>

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 tool parser

Signed-off-by: 许文卿 <xwq391974@alibaba-inc.com>

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 tool parser

Signed-off-by: 许文卿 <xwq391974@alibaba-inc.com>

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 tool parser

Signed-off-by: 许文卿 <xwq391974@alibaba-inc.com>

Xu-Wenqing committed:
Merge branch 'main' into dev/dpsk_r1_tool_parser

aarnphm reviewed (approved):
qq. The parser looks good for merging for now, but do we need toe tool chat template?

iirc this is already included in the default chat templates from the model repo?

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 tool parser

Signed-off-by: 许文卿 <xwq391974@alibaba-inc.com>

Xu-Wenqing commented:
> qq. The parser looks good for merging for now, but do we need toe tool chat template?
> 
> iirc this is already included in the default chat templates from the model repo?

yeah, DeepSeek-V3-0324 default chat template not works for vllm, fixed some issues, e.g.
`tool['function']['arguments']` changed to `tool['function']['arguments']|tojson` https://github.com/Xu-Wenqing/vllm/blob/dev/dpsk_r1_tool_parser/examples/tool_chat_template_deepseekv3.jinja#L56-L62

I updated the code, need review again, thanks.

Xu-Wenqing commented:
> Could you lint and attach a test plan?

updated descriptions, including some test cases.

aarnphm commented on vllm/entrypoints/openai/tool_parsers/deepseekv3_tool_parser.py:
```diff
@@ -0,0 +1,368 @@
+logger = init_logger(__name__)
+
+
+@ToolParserManager.register_module("deepseekv3")
```
```suggestion
@ToolParserManager.register_module("deepseek_v3")
```

minor `s/deepseekv3/deepseek_v3`

Xu-Wenqing replied:
Done.

aarnphm reviewed (approved):
Tiny

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 Function Call

Signed-off-by: Xu Wenqing <xuwq1993@qq.com>

Xu-Wenqing commented:
> Tiny

@aarnphm done.

Xu-Wenqing commented:
@DarkLight1337 need reviewers with write access, thanks.

Xu-Wenqing committed:
[Feature] Support DeepSeekV3 Function Call

Signed-off-by: Xu Wenqing <xuwq1993@qq.com>

Xu-Wenqing committed:
Merge branch 'main' into dev/dpsk_r1_tool_parser

WangErXiao commented:
@Xu-Wenqing  Can this support deepseek_r1 tool calling?

Xu-Wenqing commented:
> @Xu-Wenqing Can this support deepseek_r1 tool calling?

@WangErXiao  Yes. DeepSeek just released DeepSeek-R1-0528 model, this model support function call. We can use "deepseek_v3" tool call parser, but the chat template "tool_chat_template_deepseekv3.jinja" here don't support DeepSeek-R1-0528. I created a PR: https://github.com/vllm-project/vllm/pull/18874, add DeepSeek-R1-0528 chat template.

WangErXiao commented:
LGTM 👍 @Xu-Wenqing

xueshuai0922 commented:
@Xu-Wenqing  https://docs.vllm.ai/en/stable/features/tool_calling.html?h=function+call#deepseek-v3-models-deepseek_v3 Description in the official documentation. If the vllm version is greater than 0.8.3, it supports function call. Is there any difference between this update?

chendave reviewed (commented):
LGTM
